globus_automate_client/cli/auth.py

```diff
@@ -0,0 +1,333 @@
+            with open(self.token_store, "w") as f:
+                json.dump(
+                    {k: v._asdict() for k, v in self.tokens.items()},
+                    f,
+                    indent=2,
+                    sort_keys=True,
+                    default=default,
+                )
+        self.modified = False
+
+    def update_from_oauth_token_response(
+        self, token_response: OAuthTokenResponse, original_scopes: Set[str]
+    ) -> Dict[str, TokenSet]:
+        by_scopes = token_response.by_scopes
+        token_sets: Dict[str, TokenSet] = {}
+        for scope in by_scopes:
+            token_info = by_scopes[scope]
+            dependent_scopes = set(s for s in original_scopes if "[" in s)
+            token_set = TokenSet(
+                access_token=token_info.get("access_token"),
+                refresh_token=token_info.get("refresh_token"),
+                expiration_time=token_info.get("expires_at_seconds"),
+                dependent_scopes=dependent_scopes,
+            )
+            self.set_tokens(scope, token_set)
+            token_sets[scope] = token_set
+        self.save_tokens()
+        return token_sets
+
+
+def _get_globus_sdk_native_client(
+    client_id: str = CLIENT_ID,
+    client_name: str = CLIENT_NAME,
+):
+    return NativeAppAuthClient(client_id, app_name=client_name)
+
+
+def safeprint(s):
+    try:
+        print(s)
+        sys.stdout.flush()
+    except IOError:
+        pass
+
+
+def _do_login_for_scopes(
+    native_client: NativeAppAuthClient, scopes: List[str]
+) -> OAuthTokenResponse:
+    label = CLIENT_NAME
+    host = platform.node()
+    if host:
+        label = label + f" on {host}"
+    native_client.oauth2_start_flow(
+        requested_scopes=scopes,
+        refresh_tokens=True,
+        prefill_named_grant=label,
+    )
+    linkprompt = "Please log into Globus here"
+    safeprint(
+        "{0}:\n{1}\n{2}\n{1}\n".format(
+            linkprompt, "-" * len(linkprompt), native_client.oauth2_get_authorize_url()
+        )
+    )
+    auth_code = click.prompt("Enter the resulting Authorization Code here").strip()
+    return native_client.oauth2_exchange_code_for_tokens(auth_code)
+
+
+def get_authorizers_for_scopes(
+    scopes: List[str],
+    token_store: Optional[str] = None,
+    client_id: str = CLIENT_ID,
+    client_name: str = CLIENT_NAME,
+    no_login: bool = False,
+) -> Dict[str, GlobusAuthorizer]:
+    token_store = token_store or str(DEFAULT_TOKEN_FILE)
+    token_cache = TokenCache(token_store)
+    token_cache.load_tokens()
+    token_sets: Dict[str, TokenSet] = {}
+    needed_scopes: Set[str] = set()
+    native_client = _get_globus_sdk_native_client(client_id, client_name)
+
+    for scope in scopes:
+        token_set = token_cache.get_tokens(scope)
+        if token_set is not None:
+            token_sets[scope] = token_set
+        else:
+            needed_scopes.add(scope)
+
+    if len(needed_scopes) > 0 and not no_login:
```

Comment:
What happens if we need scopes, but no_login = False. I guess the return would never contain the authorizer for the scope, I'm just wondering if that's how it should be handled.

Reply:
yeah, I think the intention here is that in this case, we'll try to get something from the tokens we loaded from cache. In all likelihood, that will not find anything, and we just end up returning the empty `authorizers` dict.